From f79868d2adb2971fcd0327a7c01c533ea33f300d Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Fri, 8 Jun 2018 16:52:44 +0800
Subject: [PATCH 4/8] MMFMWK-8030 waylandsink: support video crop meta using
 viewporter protocol

use API wp_viewport_set_source() to support video crop meta handle

upstream status: Pending
https://bugzilla.gnome.org/show_bug.cgi?id=796541

Conflicts:
	ext/wayland/wlwindow.h
(cherry picked from commit 90ad8baff535d8c94f272f8d912c0802d5e5822b)
---
 ext/wayland/gstwaylandsink.c |  3 +++
 ext/wayland/wlwindow.c       | 30 ++++++++++++++++++++++++++++++
 ext/wayland/wlwindow.h       |  4 ++++
 3 files changed, 37 insertions(+)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 4410d1060..09e77d4ab 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -653,6 +653,7 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
   alloc = gst_wl_shm_allocator_get ();
   gst_query_add_allocation_param (query, alloc, NULL);
   gst_query_add_allocation_meta (query, GST_VIDEO_META_API_TYPE, NULL);
+  gst_query_add_allocation_meta (query, GST_VIDEO_CROP_META_API_TYPE, NULL);
   g_object_unref (alloc);
 
   return TRUE;
@@ -754,6 +755,8 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
   /* make sure that the application has called set_render_rectangle() */
   if (G_UNLIKELY (sink->window->render_rectangle.w == 0))
     goto no_window_size;
+  
+  gst_wl_window_set_source_crop (sink->window, buffer);
 
   wlbuffer = gst_buffer_get_wl_buffer (sink->display, buffer);
 
diff --git a/ext/wayland/wlwindow.c b/ext/wayland/wlwindow.c
index 66df0fce7..0a3ebf0c2 100644
--- a/ext/wayland/wlwindow.c
+++ b/ext/wayland/wlwindow.c
@@ -152,6 +152,10 @@ gst_wl_window_init (GstWlWindow * self)
   self->configured = TRUE;
   g_cond_init (&self->configure_cond);
   g_mutex_init (&self->configure_mutex);
+  self->src_x = 0;
+  self->src_y = 0;
+  self->src_width = -1;
+  self->src_height = 0;
 }
 
 static void
@@ -396,6 +400,11 @@ gst_wl_window_resize_video_surface (GstWlWindow * window, gboolean commit)
   GstVideoRectangle dst = { 0, };
   GstVideoRectangle res;
 
+  wl_fixed_t src_x = wl_fixed_from_int (window->src_x);
+  wl_fixed_t src_y = wl_fixed_from_int (window->src_y);
+  wl_fixed_t src_width = wl_fixed_from_int (window->src_width);
+  wl_fixed_t src_height = wl_fixed_from_int (window->src_height);
+
   /* center the video_subsurface inside area_subsurface */
   src.w = window->video_width;
   src.h = window->video_height;
@@ -405,6 +414,9 @@ gst_wl_window_resize_video_surface (GstWlWindow * window, gboolean commit)
   if (window->video_viewport) {
     gst_video_sink_center_rect (src, dst, &res, TRUE);
     wp_viewport_set_destination (window->video_viewport, res.w, res.h);
+    if (src_width != wl_fixed_from_int(-1))
+      wp_viewport_set_source (window->video_viewport,
+          src_x, src_y, src_width, src_height);
   } else {
     gst_video_sink_center_rect (src, dst, &res, FALSE);
   }
@@ -572,3 +584,21 @@ gst_wl_window_set_render_rectangle (GstWlWindow * window, gint x, gint y,
   if (window->video_width != 0)
     wl_subsurface_set_desync (window->video_subsurface);
 }
+
+void
+gst_wl_window_set_source_crop (GstWlWindow * window, GstBuffer * buffer)
+{
+  GstVideoCropMeta *crop = NULL;
+  crop = gst_buffer_get_video_crop_meta(buffer);
+
+  if (crop) {
+    GST_DEBUG ("buffer crop x=%d y=%d width=%d height=%d\n",
+        crop->x, crop->y, crop->width, crop->height);
+    window->src_x = crop->x;
+    window->src_y = crop->y;
+    window->src_width = crop->width;
+    window->src_height = crop->height;
+  } else {
+    window->src_width = -1;
+  }
+}
diff --git a/ext/wayland/wlwindow.h b/ext/wayland/wlwindow.h
index 303c336dd..8fa79c541 100644
--- a/ext/wayland/wlwindow.h
+++ b/ext/wayland/wlwindow.h
@@ -71,6 +71,9 @@ struct _GstWlWindow
   /* when this is not set both the area_surface and the video_surface are not
    * visible and certain steps should be skipped */
   gboolean is_area_surface_mapped;
+
+  /* the coordinate of video crop */
+  gint src_x, src_y, src_width, src_height;
 };
 
 struct _GstWlWindowClass
@@ -95,6 +98,7 @@ void gst_wl_window_render (GstWlWindow * window, GstWlBuffer * buffer,
         const GstVideoInfo * info);
 void gst_wl_window_set_render_rectangle (GstWlWindow * window, gint x, gint y,
         gint w, gint h);
+void gst_wl_window_set_source_crop (GstWlWindow * window, GstBuffer * buffer);
 
 G_END_DECLS
 
-- 
2.34.1

